Pricipal and User Resolver

kendallmiller opened the pull request:
This set of changes introduces a new IPrincipalResolver and IApplicationUserResolver, along with both default implementations of them and ones for ASP.NET Core.
The main goal is to enable accurate recording of the authenticated user of an ASP.NET web site and make it easy for individual applications to enhance how this is recorded, replacing the ResolveApplicationUser event from the previous Loupe agent.

My main concerns are:
1. Terminology:  I'm not sure my terminology is both consistent with .NET Core and obvious to users, in particular the use of "Register*" and "Unregister*".
2. Should we really be resolving the IPrincipal or the IIdentity?  I felt the former since it would include details about how the identity was being leveraged by the application.

markrendle commented on src/Core/Monitor/DefaultPrincipalResolver.cs:
```diff
@@ -0,0 +1,49 @@
+        }
+
+        /// <inheritdoc />
+        public IPrincipal ResolveCurrentPrincipal()
```
I'd probably implement this as a `public bool TryResolveCurrentPrincipal(out IPrincipal principal)` if not having one is OK, in the interest of being explicit about `null`s.

Or have an `UnknownPrincipal` type with a single value that can be used wherever an actual `IPrincipal` can't be resolved.

kendallmiller replied:
Good idea, that also makes it clear that no exception will be thrown if it can't be resolved.

kendallmiller replied:
Done.

daveaglick commented on src/Core/Monitor/IApplicationUserResolver.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// Implemented to translate an IPrincipal to a Loupe ApplicationUser
+    /// </summary>
+    public interface IApplicationUserResolver
```
IMO this isn’t really a “resolver” since you’re providing an `IPrincipal` already obtained out-of-band as input to the `ResolveApplicationUser()` call. I’d consider renaming to something like `IApplicationPrincipalAdapter` or maybe `...Provider`.

I also couldn’t find an implementation - more out of curiosity than anything else, where is it?

kendallmiller replied:
Provider sounds good - because the idea is that you inject your code to take an IPrinciple and record a much more detailed object from it (an "ApplicationUser").  

I didn't create a generic ASP.NET Core implementation yet; each app would have to make its own based on the claims and structure of its own user storage (which is what we required previously as well, just through a different API) but possibly we could do better if there's a standard "shape" to most ASP.NET Core authentication.  You can see two example implementations in the unit tests for ApplicationUserTests, e.g.:

```cs
        private class ResolveUserForCurrentPrincipal : IApplicationUserResolver
        {
            public ApplicationUser ResolveApplicationUser(IPrincipal principal, Func<ApplicationUser> userFactory)
            {
                var identity = principal.Identity;
                var newUser = userFactory();
                newUser.Caption = Environment.GetEnvironmentVariable("USERNAME");
                newUser.Organization = "Unit test";
                newUser.EmailAddress = "support@gibraltarsoftware.com";
                newUser.Phone = "443 738-0680";
                newUser.Properties.Add("Customer Key", "1234-5678-90");
                newUser.Properties.Add("License Check", null);

                return newUser;
            }
        }
```
This implementation is clearly bogus and for a unit test; you wouldn't pull the user name from the environment, you wouldn't have everything be hard coded, etc.   

What do you think of the user factory approach here?  I'm not entirely thrilled with it but it does hit the goal of providing a call to get a "blank" ApplicationUser (that under the covers has the right key associated with it and is safe for the system to serialize).

kendallmiller replied:
Interface renamed, changed method to Try* signature to match the IPrincipalProvider change and then switched to a Lazy<T> to eliminate the factory. Let me know what you think!

daveaglick replied:
Reads much better to me with these names and also really like the `Try...` suggestion from Mark. Looking good.

pcorkum commented:
It would be great to be able to handle the mapping between an IPrincipal and the Application user using a delegate instead of having to create a class that implements IApplicationUserProvider.

pcorkum commented:
Note that there is an expectation that the Name property on IIdentity will be set. The problem is that Name may not always be set. It would be wise to either document this assumption and show how to ensure it is set (for instance when the standard name claim is not available), or allow the user to configure the key for Loupe's purposes.

kendallmiller commented:
> It would be great to be able to handle the mapping between an IPrincipal and the Application user using a delegate instead of having to create a class that implements IApplicationUserProvider.

Done, added delegate versions for IPrincipalResolver, IApplicationUserProvider, and ILoupeFilter.  Also added builder overrides to make it easier to specify them as part of setting up ASP.NET Core.
